src/sentry/rules/history/preview.py

```diff
@@ -387,68 +387,92 @@ def apply_frequency_conditions(
 
+    filtered_activity = defaultdict(list)
     if condition_match == "all":
-        filtered_activity = {}
         for group, activities in group_activity.items():
-            # TODO: only EventFrequencyCondition is supported right now, so we can reuse the same query
-            buckets = get_frequency_buckets(
-                project, start, end, group, dataset_map.get(group, None)
-            )
-            if not has_issue_state_condition:
-                # If there are no issue state change conditions, then we won't have any initial activities to base our
-                # frequency condition queries off of. Instead, we take the first frequency condition and create the
-                # initial activities from that
-                for bucket_time in buckets.keys():
-                    activity = ConditionActivity(
+            init_activities_from_freq_cond = False
+            for conditions in condition_types.values():
+                # reuse frequency buckets for conditions of the same type
+                try:
+                    buckets = get_frequency_buckets(
+                        project,
+                        start,
+                        end,
                         group,
-                        ConditionActivityType.FREQUENCY_CONDITION,
-                        bucket_time,
+                        dataset_map[group],
+                        conditions[0].get_preview_aggregate(),
                     )
-                    try:
-                        if conditions[0].passes_activity_frequency(activity, buckets):
-                            activities.append(activity)
-                    except NotImplementedError:
-                        raise PreviewException
-
-            passes = [True] * len(activities)
-            for condition in conditions if has_issue_state_condition else conditions[1:]:
-                for i, activity in enumerate(activities):
-                    try:
-                        if passes[i] and not condition.passes_activity_frequency(activity, buckets):
-                            passes[i] = False
-                    except NotImplementedError:
-                        raise PreviewException
-
-            filtered_activity[group] = [activities[i] for i in range(len(activities)) if passes[i]]
-
-        return filtered_activity
+                except NotImplementedError:
+                    raise PreviewException
+                skip_first = False
+                if not has_issue_state_condition and not init_activities_from_freq_cond:
+                    # If there are no issue state change conditions, then we won't have any initial activities
+                    # to base our frequency condition queries off of. Instead, we take the first frequency condition and
+                    # create the initial activities from that
+                    init_activities_from_freq_cond = skip_first = True
```

Comment:
what's the purpose of setting these statically if they don't get modified later?

Reply:
These flags are initialized are both initialized to false, and then set to true here because we want to either skip the first condition (`skip_first`) or not enter this if block again for this group (`init_activities_from_freq_cond`)

Reply:
before we could just iterate over all the conditions at once, but now that we separate them by id we have to first iterate over `condition_types` and then all the `conditions` in a type

Reply:
ah i see. so if we don't have any groups we use the first one to create the initial query